subprojects/maven/src/main/java/org/gradle/api/publish/maven/tasks/PublishToMavenRepository.java

```diff
@@ -104,6 +129,18 @@ private PublishSpec computeSpec() {
         );
     }
 
+    private boolean areCredentialsSafe(String identity, Credentials toCheck) {
+        ProviderFactory providerFactory = getServices().get(ProviderFactory.class);
+        Credentials referenceCredentials;
+        try {
+            Provider<? extends Credentials> credentialsProvider = providerFactory.credentials(toCheck.getClass(), identity);
```

Comment:
🤔 I wonder if the concrete `Credentials` type at this point could be a generated type. In that case `unpackType` would make sense (`org.gradle.api.internal.GeneratedSubclasses#unpackType`).

Reply:
> 🤔 I wonder if the concrete `Credentials` type at this point could be a generated type. In that case `unpackType` would make sense (`org.gradle.api.internal.GeneratedSubclasses#unpackType`).

Are they subclasses of the base type (like the `_Decorated` classes)? I saw them indeed, and was concerned at first, but in that case, as long as the ProviderFactory.credentials() is using instanceOf, we are good, no?

Reply:
In that case it would be safer to use `unpackType` and only pass around / store the unpacked type.